# 1. Polars

In [1]:
%pip install polars

In [2]:
import polars as pl

Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью polars - 1 балл

In [10]:
df = pl.read_csv("train.csv")

Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д. - 1 балл
hint: используйте методы и атрибуты датафрейма: describe(), dtypes, null_count(), mean()...

In [5]:
print("Типы данных:", df.dtypes)
print("Количество пропусков:", df.null_count())
print("Описание данных:\n", df.describe())
print("Средние значения:\n", df.select([pl.col(column).mean() for column in df.columns if df.schema[column] == pl.Float64]))

Типы данных: [Int64, Int64, Int64, String, String, Float64, Int64, Int64, String, Float64, String, String]
Количество пропусков: shape: (1, 12)
┌─────────────┬──────────┬────────┬──────┬───┬────────┬──────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name ┆ … ┆ Ticket ┆ Fare ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---  ┆   ┆ ---    ┆ ---  ┆ ---   ┆ ---      │
│ u32         ┆ u32      ┆ u32    ┆ u32  ┆   ┆ u32    ┆ u32  ┆ u32   ┆ u32      │
╞═════════════╪══════════╪════════╪══════╪═══╪════════╪══════╪═══════╪══════════╡
│ 0           ┆ 0        ┆ 0      ┆ 0    ┆ … ┆ 0      ┆ 0    ┆ 687   ┆ 2        │
└─────────────┴──────────┴────────┴──────┴───┴────────┴──────┴───────┴──────────┘
Описание данных:
 shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   


Посчитайте количество пассажиров каждого класса (Pclass) - 1 балл
hint: используйте методы: get_column(), value_counts()...

In [13]:
df.get_column("Pclass").value_counts()

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 1      ┆ 216   │
│ 2      ┆ 184   │
│ 3      ┆ 491   │
└────────┴───────┘



Выведите количество выживших мужчин и женщин на корабле - 1 балл
hint: используйте методы: groupby(), agg(), pl.col(), sum()...

In [14]:
survived_by_gender = (
    df.group_by("Sex")
    .agg(pl.col("Survived").sum().alias("Survived_Count"))
)
survived_by_gender

shape: (2, 2)
┌────────┬────────────────┐
│ Sex    ┆ Survived_Count │
│ ---    ┆ ---            │
│ str    ┆ i64            │
╞════════╪════════════════╡
│ male   ┆ 109            │
│ female ┆ 233            │
└────────┴────────────────┘



Выведите часть таблицы с пассажирами, возраст которых больше 44 лет - 1 балл
hint: используйте методы: filter(), pl.col()...

In [15]:
older_passengers = df.filter(pl.col("Age") > 44)
older_passengers

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0    ┆ null  ┆ S        │
│             ┆          ┆        ┆ (Mary D          ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Kingcome…        ┆   ┆            ┆         ┆       ┆          │
│ 34          ┆ 0        ┆ 2      ┆ Wheadon, Mr.     ┆ … ┆ C.A. 24579 ┆ 10.5    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Edward H         ┆   ┆            ┆         ┆       ┆          │
│ 53          ┆ 1        ┆ 1      ┆ Harper, Mrs.     ┆ … ┆ PC 17572   ┆ 76.7292 ┆ D33   ┆ C        │
│             ┆          ┆        ┆ Henry Sleeper    ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (My…             ┆   ┆            ┆         ┆       ┆          │
│ …           ┆ …        ┆ …      ┆ …                ┆ … ┆ …          ┆ …       ┆ …     ┆ …        │
│ 858         ┆ 1        ┆ 1      ┆ Daly, Mr. Peter  ┆ … ┆ 113055     ┆ 26.55   ┆ E17   ┆ S        │
│             ┆          ┆        ┆ Denis            ┆   ┆            ┆         ┆       ┆          │
│ 863         ┆ 1        ┆ 1      ┆ Swift, Mrs.      ┆ … ┆ 17466      ┆ 25.9292 ┆ D17   ┆ S        │
│             ┆          ┆        ┆ Frederick Joel   ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Ma…             ┆   ┆            ┆         ┆       ┆          │
│ 872         ┆ 1        ┆ 1      ┆ Beckwith, Mrs.   ┆ … ┆ 11751      ┆ 52.5542 ┆ D35   ┆ S        │
│             ┆          ┆        ┆ Richard Leonard… ┆   ┆            ┆         ┆       ┆          │
│ 874         ┆ 0        ┆ 3      ┆ Vander Cruyssen, ┆ … ┆ 345765     ┆ 9.0     ┆ null  ┆ S        │
│             ┆          ┆        ┆ Mr. Victor       ┆   ┆            ┆         ┆       ┆          │
│ 880         ┆ 1        ┆ 1      ┆ Potter, Mrs.     ┆ … ┆ 11767      ┆ 83.1583 ┆ C50   ┆ C        │
│             ┆          ┆        ┆ Thomas Jr (Lily  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ A…               ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

# Ускорение работы с pandas

In [16]:
%pip install bottleneck

In [17]:
import pandas as pd
import bottleneck as bn

Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью pandas


In [18]:
df = pd.read_csv("train.csv")


Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck - 1 бал
hint: используйте методы: bn.nanmean(), bn.nanstd()


In [20]:
mean_age = bn.nanmean(df["Age"])
std_age = bn.nanstd(df["Age"])
print("Средний возраст пассажиров:", mean_age)
print("Стандартное отклонение:", std_age)

Средний возраст пассажиров: 29.69911764705882
Стандартное отклонение: 14.516321150817317



Для каждого пассажира умножьте значение столбца Fare на 1.3 и сохраните результаты как новый столбец Fare_new - 1 балл
hint: используйте методы: .itertuples() или apply()

In [23]:
df["Fare_new"] = df["Fare"].apply(lambda x: x * 1.3)
df[["Fare", "Fare_new"]].head()

Fare  Fare_new
0   7.2500   9.42500
1  71.2833  92.66829
2   7.9250  10.30250
3  53.1000  69.03000
4   8.0500  10.46500

In [24]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Fare_new  
0      0         A/5 21171   7.2500   NaN        S   9.42500  
1      0          PC 17599  71.2833   C85        C  92.66829  
2      0  STON/O2. 3101282   7.9250   NaN        S  10.30250  
3      0            113803  53.1000  C123        S  69.03000  
4      0            373450   8.0500   NaN        S  10.46500

# Оптимизация типов pandas

Считайте датасет из файла Housing.csv (это данные о ценах домов) с помощью pandas

In [25]:
df = pd.read_csv("Housing.csv")


Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях - 2 балла


Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации - 1 балл


In [26]:
initial_memory_usage = df.memory_usage(deep=True).sum() / 1024 ** 2

In [28]:
optimized_types = {}
for column in df.columns:
    col_data = df[column]
    if pd.api.types.is_integer_dtype(col_data):
        optimized_types[column] = pd.to_numeric(col_data, downcast="integer").dtype
    elif pd.api.types.is_float_dtype(col_data):
        optimized_types[column] = "float32"
    elif pd.api.types.is_object_dtype(col_data):
        if col_data.nunique() / len(col_data) < 0.5:
            optimized_types[column] = "category"

In [29]:
df_optimized = df.astype(optimized_types)

In [30]:
optimized_memory_usage = df_optimized.memory_usage(deep=True).sum() / 1024 ** 2  # в MB

In [31]:
print("Исходное потребление памяти:", initial_memory_usage, "MB")
print("Потребление памяти после оптимизации:", optimized_memory_usage, "MB")
print("Оптимизированные типы данных:", optimized_types)

Исходное потребление памяти: 0.245819091796875 MB
Потребление памяти после оптимизации: 0.010555267333984375 MB
Оптимизированные типы данных: {'price': dtype('int32'), 'area': dtype('int16'), 'bedrooms': dtype('int8'), 'bathrooms': dtype('int8'), 'stories': dtype('int8'), 'mainroad': 'category', 'guestroom': 'category', 'basement': 'category', 'hotwaterheating': 'category', 'airconditioning': 'category', 'parking': dtype('int8'), 'prefarea': 'category', 'furnishingstatus': 'category'}


Вывод: применение оптимизированных типов данных позволило сильно снизить потребление памяти (с 0.246 МБ до 0.011 МБ), благодаря более компактным типам для числовых столбцов и замене текстовых данных на категориальные